gabbi/httpclient.py

```diff
@@ -11,53 +11,58 @@
 # License for the specific language governing permissions and limitations
 # under the License.
 
+import logging
 import os
 import sys
 
-import certifi
-import urllib3
+import httpx
 
 from gabbi.handlers import jsonhandler
 from gabbi import utils
 
-# Disable SSL warnings otherwise tests which process stderr will get
-# extra information.
-urllib3.disable_warnings()
+logging.getLogger('httpx').setLevel(logging.WARNING)
 
 
-class Http(urllib3.PoolManager):
-    """A subclass of the ``urllib3.PoolManager`` to munge the data.
+class Http:
+    """A class to munge the HTTP response.
 
     This transforms the response to look more like what httplib2
     provided when it was used as the HTTP client.
     """
 
+    def __init__(self, **kwargs):
+        self.extensions = {}
+        if 'server_hostname' in kwargs:
+            self.extensions['sni_hostname'] = kwargs['server_hostname']
+        self.client = httpx.Client(verify=kwargs.get('cert_validate', True))
+
     def request(self, absolute_uri, method, body, headers, redirect, timeout):
-        if redirect:
-            retry = urllib3.util.Retry(raise_on_redirect=False, redirect=5)
-        else:
-            retry = urllib3.util.Retry(total=False, redirect=False)
-        response = super(Http, self).request(
-            method,
-            absolute_uri,
-            body=body,
-            headers=headers,
-            retries=retry,
-            timeout=timeout,
-        )
+        try:
+            response = self.client.request(
+                method=method,
+                url=absolute_uri,
+                headers=headers,
+                content=body,
+                timeout=timeout,
+                follow_redirects=redirect,
+                extensions=self.extensions,
+            )
+        except httpx.ConnectError as error:
```

Comment:
What's the goal behind re-raising this? What happens if you just let it raise?

Reply:
Ahh, this was a little bit of debugging.  I wanted to include the method and URL in the exception string (as, by default, it just says, `invalid hostname` (or something similar)).  We can remove it, if needs be.

Reply:
Augmenting the exception with more data is potentially useful but turning it into a RuntimeError is probably not the best choice, might be better to have a custom exception within gabbi that contains the other exception.

The other option is making sure that wherever the exception is eventually caught, somewhere else in gabbi, has more sufficient messaging to provide the necessary context for which test did the failing.

Generally the pattern you want to follow is to not mess with exception this deep, and let them rise up somewhere  else that catches them and makes them meaningful to the observer.

Reply:
Absolutely.  To be honest, I was trying to take on the replacement of the WSGI-Intercept stuff too and this was added for that.  I've removed it.